### PR TITLE
[SYCL] Fix segmentation fault in program manager

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch fixes sporadic segmentation fault caused by data race in ```NativePrograms``` variable of ```ProgramManager``` class.  
According to documentation, an access to this variable should be synchronized via the same lock as program cache but there was a lack of lock guard for ```NativePrograms[Res] = &Img;``` in sycl/source/detail/program_manager/program_manager.cpp. Added missing lock guard.

### CODE REVIEW

**Comment by kbobrovs** [comment]
@dm-vodopyanov , please provide some details of the problem and the fix in the PR description and commit message.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -288,7 +288,7 @@ static const char *getFormatStr(RT::PiDeviceBinaryType Format) {
 }
 
 RT::PiProgram ProgramManager::createPIProgram(const RTDeviceBinaryImage &Img,
-                                              const context &Context) {
+                                              const context &Context) {
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp]: unneeded change

Comment (dm-vodopyanov) [reply sycl/source/detail/program_manager/program_manager.cpp]: I don't know why it happened but I already removed it in the commit above

**Comment by dm-vodopyanov** [comment]
@kbobrovs, added description, please review.

**Pull request merged**